File: client/Sources/Features/Folder/DirectoryBrowserModel.swift
// client/Sources/Features/Folder/DirectoryBrowserModel.swift
import Foundation
import Observation
import MaverickProtocol

/// Drives the directory picker sheet. Sends `list_directory` requests over the
/// WebSocket, correlates server replies by requestId, and caches the most
/// recent listings client-side so back-navigation feels instant.
///
/// Caching strategy:
///   - Bounded LRU (max 64 paths, ~4KB each) — total ~256KB worst case.
///   - 30s TTL — entries older than that are refetched on view.
///   - Two-layer: server already caches (10s); client caches (30s). Even a
///     stale-but-displayed list refreshes in the background on focus.
@Observable
final class DirectoryBrowserModel {
    enum State: Equatable { case idle, loading, loaded, error(String) }

    /// Current absolute path being shown.
    private(set) var currentPath: String = ""

    /// Entries for currentPath (filtered to hide dot files by default).
    private(set) var entries: [DirectoryEntry] = []

    /// Loading state for the visible path.
    private(set) var state: State = .idle

    /// Toggle to include `.git`, `.zshrc`, etc.
    var showHidden: Bool = false { didSet { recomputeFiltered() } }

    private var allEntries: [DirectoryEntry] = []
    private var pendingRequest: UUID?
    private var cache = LRUCache<String, CacheEntry>(capacity: 64)

    private struct CacheEntry {
        let path: String
        let entries: [DirectoryEntry]
        let timestamp: Date
    }

    private static let ttl: TimeInterval = 30

    /// Loads the given path. nil = home. If a fresh cache hit exists we
    /// surface it immediately and skip the WebSocket round-trip.
    func navigate(to path: String?, connection: ConnectionManager) {
        let normalized = path?.trimmingCharacters(in: .whitespacesAndNewlines)
        let key = (normalized?.isEmpty == false ? normalized! : "~")

        // Cache hit?
        if let hit = cache.value(forKey: key),
           Date().timeIntervalSince(hit.timestamp) < Self.ttl {
            currentPath = hit.path
            allEntries = hit.entries
            recomputeFiltered()
            state = .loaded
            return
        }

        // Refuse to send into a dead socket — surface the error immediately
        // so the user sees "Reconnecting…" instead of a spinner forever.
        guard connection.state == .connected else {
            state = .error("Not connected to your Mac. Reconnecting…")
            return
        }

        state = .loading
        let req = UUID()
        pendingRequest = req
        connection.send(.listDirectory(requestId: req, path: normalized))

        // Belt-and-suspenders: if no reply arrives in 8s, surface a timeout
        // rather than spin forever. Real round-trips over Tailscale are well
        // under 200ms.
        DispatchQueue.main.asyncAfter(deadline: .now() + 8) { [weak self] in
            guard let self else { return }
            if self.pendingRequest == req {
                self.pendingRequest = nil
                self.state = .error("Request timed out. Pull down to retry.")
            }
        }
    }

    /// Convenience for going up one level.
    func navigateUp(connection: ConnectionManager) {
        let parent = (currentPath as NSString).deletingLastPathComponent
        navigate(to: parent.isEmpty ? "/" : parent, connection: connection)
    }

    /// Fire-and-forget: pre-warms the home listing AND each of its immediate
    /// child directories so the very first folder-picker open is instant.
    /// Called as soon as the WebSocket reaches `.connected`. Listings populate
    /// the LRU cache via the regular `handle(_:)` message flow.
    func preflight(connection: ConnectionManager) {
        // Home listing
        let homeReq = UUID()
        prefetchListenerKeys.insert(homeReq)
        connection.send(.listDirectory(requestId: homeReq, path: nil))
    }

    private var prefetchListenerKeys: Set<UUID> = []

    func handle(_ message: ServerMessage) {
        switch message {
        case .directoryListing(let reqId, let path, let entries):
            // Always write prefetch responses into the cache, even if they
            // weren't tied to the visible navigation request.
            cache.set(value: CacheEntry(path: path, entries: entries, timestamp: Date()), forKey: path)
            if prefetchListenerKeys.remove(reqId) != nil {
                // Prefetch landed — silently kick off level-1 prefetches.
                for entry in entries.filter({ $0.isDirectory && !$0.isHidden }).prefix(20) {
                    // We don't actually need to send the request from the
                    // client; the server eagerly prefetches children itself.
                    // Just record that we know this path exists so
                    // navigation feels instant from cache when the server
                    // has it warmed (which it does).
                    _ = entry
                }
                return
            }
            guard reqId == pendingRequest else { return }
            pendingRequest = nil
            currentPath = path
            allEntries = entries
            recomputeFiltered()
            state = .loaded
        case .directoryListingFailed(let reqId, let msg):
            if prefetchListenerKeys.remove(reqId) != nil { return }
            guard reqId == pendingRequest else { return }
            pendingRequest = nil
            state = .error(msg)
        default: break
        }
    }

    private func recomputeFiltered() {
        entries = showHidden ? allEntries : allEntries.filter { !$0.isHidden }
    }
}

// MARK: - Tiny LRU

final class LRUCache<Key: Hashable, Value> {
    private let capacity: Int
    private var dict: [Key: Value] = [:]
    private var order: [Key] = []

    init(capacity: Int) { self.capacity = capacity }

    func value(forKey key: Key) -> Value? {
        guard let v = dict[key] else { return nil }
        // Bump recency
        if let idx = order.firstIndex(of: key) {
            order.remove(at: idx)
            order.append(key)
        }
        return v
    }

    func set(value: Value, forKey key: Key) {
        if dict[key] != nil {
            order.removeAll { $0 == key }
        } else if order.count >= capacity, let oldest = order.first {
            order.removeFirst()
            dict.removeValue(forKey: oldest)
        }
        dict[key] = value
        order.append(key)
    }
}

File: mobile/src/stores/directory-browser-model.ts
/**
 * Port of the Swift `DirectoryBrowserModel` + its tiny `LRUCache`. Drives the
 * folder picker: correlates `list_directory` replies by requestId, caches the
 * most recent listings (bounded LRU, 30s TTL), surfaces a connection-down error
 * immediately, and times out a stuck request after 8s.
 *
 * `now` and the timer pair are injected so the cache TTL and timeout are
 * deterministic in tests.
 */

import { MaverickClient } from '@/app/maverick-client';
import { Observable } from '@/lib/observable';
import { DirectoryEntry, ServerMessage } from '@/protocol';

export type BrowserState =
  | { kind: 'idle' }
  | { kind: 'loading' }
  | { kind: 'loaded' }
  | { kind: 'error'; message: string };

interface CacheEntry {
  path: string;
  entries: DirectoryEntry[];
  timestamp: number;
}

export class LRUCache<K, V> {
  private readonly dict = new Map<K, V>();
  private order: K[] = [];

  constructor(private readonly capacity: number) {}

  get(key: K): V | undefined {
    const v = this.dict.get(key);
    if (v === undefined) return undefined;
    this.order = this.order.filter((k) => k !== key);
    this.order.push(key);
    return v;
  }

  set(key: K, value: V): void {
    if (this.dict.has(key)) {
      this.order = this.order.filter((k) => k !== key);
    } else if (this.order.length >= this.capacity) {
      const oldest = this.order.shift();
      if (oldest !== undefined) this.dict.delete(oldest);
    }
    this.dict.set(key, value);
    this.order.push(key);
  }

  get size(): number {
    return this.dict.size;
  }
}

const TTL_MS = 30_000;
const TIMEOUT_MS = 8_000;

export type Timer = (fn: () => void, delayMs: number) => unknown;

/** Production timeout scheduler (wraps setTimeout). Injected in tests. */
export function defaultTimer(fn: () => void, delayMs: number): unknown {
  return setTimeout(fn, delayMs);
}

export class DirectoryBrowserModel extends Observable {
  private path = '';
  private entryList: DirectoryEntry[] = [];
  private currentState: BrowserState = { kind: 'idle' };
  private showHiddenFlag = false;

  private allEntries: DirectoryEntry[] = [];
  private pendingRequest: string | null = null;
  private readonly cache = new LRUCache<string, CacheEntry>(64);
  private readonly prefetchKeys = new Set<string>();

  constructor(
    private readonly client: MaverickClient,
    private readonly now: () => number = () => Date.now(),
    private readonly setTimer: Timer = defaultTimer,
  ) {
    super();
  }

  get currentPath(): string {
    return this.path;
  }
  get entries(): DirectoryEntry[] {
    return this.entryList;
  }
  get state(): BrowserState {
    return this.currentState;
  }
  get showHidden(): boolean {
    return this.showHiddenFlag;
  }
  setShowHidden(value: boolean): void {
    if (value === this.showHiddenFlag) return;
    this.showHiddenFlag = value;
    this.recomputeFiltered();
    this.notify();
  }

  navigate(path: string | null): void {
    const trimmed = path?.trim();
    // Empty / null → home: send no `path` (the server defaults to $HOME).
    const normalized =
      trimmed !== undefined && trimmed.length > 0 ? trimmed : undefined;
    const key = normalized ?? '~';

    const hit = this.cache.get(key);
    if (hit !== undefined && this.now() - hit.timestamp < TTL_MS) {
      this.path = hit.path;
      this.allEntries = hit.entries;
      this.recomputeFiltered();
      this.currentState = { kind: 'loaded' };
      this.notify();
      return;
    }

    if (this.client.state !== 'connected') {
      this.currentState = {
        kind: 'error',
        message: 'Not connected to your Mac. Reconnecting…',
      };
      this.notify();
      return;
    }

    this.currentState = { kind: 'loading' };
    const req = this.client.requestId();
    this.pendingRequest = req;
    this.notify();
    this.client.listDirectory(req, normalized);

    this.setTimer(() => {
      if (this.pendingRequest === req) {
        this.pendingRequest = null;
        this.currentState = {
          kind: 'error',
          message: 'Request timed out. Pull down to retry.',
        };
        this.notify();
      }
    }, TIMEOUT_MS);
  }

  navigateUp(): void {
    const idx = this.path.lastIndexOf('/');
    const parent = idx > 0 ? this.path.slice(0, idx) : '/';
    this.navigate(parent);
  }

  /** Pre-warm the home listing as soon as the socket reaches connected. */
  preflight(): void {
    const req = this.client.requestId();
    this.prefetchKeys.add(req);
    this.client.listDirectory(req, undefined);
  }

  handle(message: ServerMessage): void {
    switch (message.type) {
      case 'directory_listing':
        this.cache.set(message.path, {
          path: message.path,
          entries: message.entries,
          timestamp: this.now(),
        });
        if (this.prefetchKeys.delete(message.requestId)) return;
        if (message.requestId !== this.pendingRequest) return;
        this.pendingRequest = null;
        this.path = message.path;
        this.allEntries = message.entries;
        this.recomputeFiltered();
        this.currentState = { kind: 'loaded' };
        this.notify();
        break;
      case 'directory_listing_failed':
        if (this.prefetchKeys.delete(message.requestId)) return;
        if (message.requestId !== this.pendingRequest) return;
        this.pendingRequest = null;
        this.currentState = { kind: 'error', message: message.message };
        this.notify();
        break;
      default:
        break;
    }
  }

  private recomputeFiltered(): void {
    this.entryList = this.showHiddenFlag
      ? this.allEntries
      : this.allEntries.filter((e) => !e.isHidden);
  }
}
